gcloud-java-dns/src/main/java/com/google/gcloud/dns/DnsBatch.java

```diff
@@ -0,0 +1,351 @@
+import com.google.api.services.dns.model.ResourceRecordSet;
+import com.google.api.services.dns.model.ResourceRecordSetsListResponse;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Iterables;
+import com.google.gcloud.Page;
+import com.google.gcloud.PageImpl;
+import com.google.gcloud.dns.spi.DnsRpc;
+import com.google.gcloud.dns.spi.RpcBatch;
+
+import java.util.List;
+import java.util.Map;
+
+/**
+ * A batch of operations to be submitted to Google Cloud DNS using a single RPC request.
+ */
+public class DnsBatch {
+
+  private RpcBatch batch;
+  private final DnsRpc dnsRpc;
+  private final DnsOptions options;
+
+  DnsBatch(DnsOptions options) {
+    this.options = options;
+    this.dnsRpc = options.rpc();
+    this.batch = dnsRpc.createBatch();
+  }
+
+  Object batch() {
+    return batch;
+  }
+
+  DnsRpc dnsRpc() {
+    return dnsRpc;
+  }
+
+  DnsOptions options() {
+    return options;
+  }
+
+  /**
+   * Adds a request representing the "list zones" operation to this batch. The {@code options} can
+   * be used to restrict the fields returned or provide page size limits in the same way as for
+   * {@link Dns#listZones(Dns.ZoneListOption...)}. The returned {@link DnsBatchResult} will return a
+   * page of zones upon calling {@link DnsBatchResult#get()} on successful completion, or it will
+   * throw a {@link DnsException} if the operation failed.
+   */
+  public DnsBatchResult<Page<Zone>> listZones(Dns.ZoneListOption... options) {
+    DnsBatchResult<Page<Zone>> result = new DnsBatchResult<>();
+    Map<DnsRpc.Option, ?> optionMap = DnsImpl.optionMap(options);
+    RpcBatch.Callback<ManagedZonesListResponse> callback = createListZonesCallback(result,
+        optionMap);
+    batch.addListZones(callback, optionMap);
+    return result;
+  }
+
+  /**
+   * Adds a request representing the "create zone" operation to this batch. The {@code options} can
+   * be used to restrict the fields returned in the same way as for {@link Dns#create(ZoneInfo,
+   * Dns.ZoneOption...)}. The returned {@link DnsBatchResult} will return the created {@link Zone}
+   * upon calling {@link DnsBatchResult#get()} on successful completion, or it will throw a {@link
+   * DnsException} if the operation failed.
+   */
+  public DnsBatchResult<Zone> createZone(ZoneInfo zone, Dns.ZoneOption... options) {
+    DnsBatchResult<Zone> result = new DnsBatchResult<>();
+    RpcBatch.Callback<ManagedZone> callback = createZoneCallback(this.options, result);
+    Map<DnsRpc.Option, ?> optionMap = DnsImpl.optionMap(options);
+    batch.addCreateZone(zone.toPb(), callback, optionMap);
+    return result;
+  }
+
+  /**
+   * Adds a request representing the "delete zone" operation to this batch. The returned {@link
+   * DnsBatchResult} will return {@code true} upon calling {@link DnsBatchResult#get()} on
+   * successful deletion, {@code false} if the zone was not found, or it will throw a {@link
```

Comment:
Grammar nit: inconsistent listing here.  It could be changed to "...will return true..., return false..., or throw a..."